pkgs/build-support/node/fetch-npm-deps/src/main.rs

```diff
@@ -0,0 +1,178 @@
+#![warn(clippy::pedantic)]
+
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+use sha2::{digest::Update, Digest, Sha256};
+use std::{
+    collections::HashMap,
+    env,
+    fs::{self, File},
+    io::{self, Write},
+    path::PathBuf,
+    process::Command,
+};
+
+#[derive(Deserialize)]
+struct PackageLock {
+    #[serde(rename = "lockfileVersion")]
+    version: u8,
+    dependencies: Option<HashMap<String, OldPackage>>,
+    packages: Option<HashMap<String, Package>>,
+}
+
+#[derive(Deserialize)]
+struct OldPackage {
+    version: String,
+    resolved: Option<String>,
+    integrity: Option<String>,
+    dependencies: Option<HashMap<String, OldPackage>>,
+}
+
+#[derive(Deserialize)]
+struct Package {
+    resolved: Option<String>,
+    integrity: Option<String>,
+}
+
+#[derive(Serialize)]
+struct Key {
+    key: String,
+    integrity: String,
+    time: u8,
+    size: u64,
+    metadata: HashMap<u8, u8>,
+}
+
+fn to_new_packages(old_packages: HashMap<String, OldPackage>) -> HashMap<String, Package> {
+    let mut new = HashMap::new();
+
+    for (name, package) in old_packages {
+        if package.integrity.is_some() {
+            new.insert(
+                format!("{name}-{}", package.version),
+                Package {
+                    resolved: if package.version.starts_with("https") {
+                        Some(package.version)
+                    } else {
+                        package.resolved
+                    },
+                    integrity: package.integrity,
+                },
+            );
+
+            if let Some(dependencies) = package.dependencies {
+                new.extend(to_new_packages(dependencies));
+            }
+        }
+    }
+
+    new
+}
+
+fn push_hash_segments(path: &mut PathBuf, hash: &str) {
+    path.push(&hash[0..2]);
+    path.push(&hash[2..4]);
+    path.push(&hash[4..]);
+}
+
+fn hash_digest(data: impl AsRef<[u8]>) -> String {
+    format!("{:x}", Sha256::new().chain(data.as_ref()).finalize())
+}
+
+fn main() -> anyhow::Result<()> {
+    let mut args = env::args().skip(1);
+
+    let lock: PackageLock = serde_json::from_str(&fs::read_to_string(args.next().unwrap())?)?;
+
+    let out = PathBuf::from(args.next().unwrap()).join("_cacache");
+
+    let print_hash = args.next().is_some();
```

Comment:
Adding a help message would be good, considering this will be used by update scripts.